nntrainer/models/neuralnet.cpp

```diff
@@ -643,11 +643,17 @@ int NeuralNetwork::train_run() {
   auto &in = first_layer->net_input[0]->getVariableRef();
 
   /// @todo migrate this to trait based system; sth like need label?
-  std::shared_ptr<Layer> knn;
+  std::shared_ptr<Layer> layer_;
   for (auto &layer_node : model_graph.getSorted()) {
-    knn = layer_node.layer;
-    if (knn->getType() == "centroid_knn") {
-      knn->net_hidden[0]->getGradientRef() = label;
+    layer_ = layer_node.layer;
+    if (layer_->getType() == "centroid_knn") {
+      layer_->net_hidden[0]->getGradientRef() = label;
+    }
+
+    if (layer_->getType() == "RNN"){
+      Tensor &var_ref= layer_->net_hidden[0]->getVariableRef();
+      // for now, we are using Xavir_normal for hidden initialization
+      var_ref.setRandNormal(0.0f, sqrtFloat(2.0f / (var_ref.width() + var_ref.height())));
```

Comment:
1. `HiddenLayer value` must be a weight for RNN, and not the output. If it is an output, it can be overwritten by other layers/derivatives, but that value is needed for the next iteration. Further, during inference, its value would be overwritten due to optimization but hidden value is needed for the next iteration.
2. Initialization of hidden value must be done inside the RNN layer itself, and not by the model. `centroid_knn` does it in the model which is wrong. It requires to pass the label for the training/validation and should be done by forwarding/backwarding like done for loss layer.

Reply:
right. it removed. Thanks.